### ISSUE TITLE
Enhancement: [no-unsafe-declaration-merging] Switch to scope manager

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#5854]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/no-unsafe-declaration-merging/

### Description

Per https://github.com/typescript-eslint/typescript-eslint/pull/5840#discussion_r999330659: 

> We should actually be able to do this without type info and just with the scope analysis APIs.

Not using type information where possible is preferable for rules. See https://typescript-eslint.io/docs/linting/typed-linting/.

### Fail

```typescript
(same as before)
```

### Pass

```typescript
(same as before)
```

### Additional Info

This rule is only in the strict config, not one of the recommended ones (yet). So it wouldn't be a breaking change to use the scope manager.

**Comment by bradzacher** [comment]
It's worth noting that by switching from type information we won't be able to detect these two cases:

```ts
namespace Foo {
  export interface Bar {}
}
namespace Foo {
  export class Bar {}
}
```

```ts
import Foo from './Foo';

namespace Foo {
  export interface Bar {}
}
```

I don't think that's a huge problem given that it would vastly improve the performance and usability of the rule. Those are pretty rare edge-cases as well.
